File: scala/src/main/scala/leetcode/editor/cn/[19]删除链表的倒数第 N 个结点.scala
package leetcode.editor.cn
object RemoveNthNodeFromEndOfList {

  class ListNode(_x: Int = 0, _next: ListNode = null) {
    var next: ListNode = _next
    var x: Int = _x
  }
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for singly-linked list.
    */
  object Solution {
    def removeNthFromEnd(head: ListNode, n: Int): ListNode = {
      val dummy = ListNode(-1)
      dummy.next = head
      val x = findFromEnd(dummy, n + 1)
      x.next = x.next.next
      dummy.next
    }

    def findFromEnd(head: ListNode, k: Int): ListNode = {
      var p1 = head
      var p2 = head
      
      for (_ <- 0 until k) p1 = p1.next

      while (p1 != null) {
        p2 = p2.next
        p1 = p1.next
      }
      p2
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}

File: cpp/leetcode/editor/cn/[19]删除链表的倒数第 N 个结点.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        ListNode dummy(-1);
        dummy.next = head;

        auto x = findFromEnd(&dummy, n + 1);
        x->next = x->next->next;
        return dummy.next;
    }

    ListNode *findFromEnd(ListNode *head, int k) {
        auto p1 = head;
        for (auto i = 0; i < k; i++) p1 = p1->next;

        auto p2 = head;
        while (p1 != nullptr) {
            p2 = p2->next;
            p1 = p1->next;
        }
        return p2;
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}